### ISSUE TITLE
Option of settings being at server/deployment level?

### ISSUE THREAD

**Issue by EDIflyer** [amir20/dozzle#2437]
**Is your feature request related to a problem? Please describe.**
I use Dozzle (which is totally brilliant BTW!) on multiple different computers and browsers.  Currently I need to set all my customisations anew on each browser I'm using.

**Describe the solution you'd like**
The option for the settings to apply across all browsers - I was thinking this could be particulary useful for remembering upcoming filter settings (#2432), pinned/stored searches (#2036), etc.

**Describe alternatives you've considered**
Continue the current model

**Additional context**
I saw your helpful post in #1859 where you explained the three levels of config in Dozzle:
> There are three scopes of configurations in Dozzle:
> 1. User scope which is specific to a browser (localstorage)
> 2. Deployment scope which is per install and configured for all users
> 3. And no scope, something like hardcoded in binary

I guess what I'm asking for is the option to, on a per-deployment basis, switch to have the settings applying at level 2 rather than level 1.  Although this could be an option in the UI I wonder if it might work easier/better as an environment variable used when running the docker container (e.g., a `--server_settings` flag in docker run or `DOZZLE_SERVER_SETTINGS: true` in docker-compose) - this would get around any issues for setups where there _are_ multiple users and you wouldn't want one user overriding another's settings and obviously wouldn't change the current default of level 1.

I was envisaging this working something along the lines of any setting changes still being made to `localstorage` as normal but then also being sent to the server. When first loading the Dozzle app it would then ask if there were any settings stored on the server and use them to populate `localstorage`.  If you made a setting change on one browser then a manual refresh on the other browsers would be required to reload the settings.

If this is too big/complicated/niche a request don't worry - it just would save me having to remember to change a setting across lots of browsers and as mentioned above I was thinking it might become even more handy going forward if users start to have more by way of customisations.

Hope the above makes sense and thanks again for all your work on this!

**Comment by amir20** [comment]
Thanks! I have been thinking about this a lot. There are two parts: 

Technically, I was thinking just storing everything as json. I also thought about using some kind of embedded database but I think that's too complicated. 

The harder part is configuration. 

> I guess what I'm asking for is the option to, on a per-deployment basis, switch to have the settings applying at level 2 rather than level 1. Although this could be an option in the UI I wonder if it might work easier/better as an environment variable used when running the docker container (e.g., a --server_settings flag in docker run or DOZZLE_SERVER_SETTINGS: true in docker-compose) - this would get around any issues for setups where there are multiple users and you wouldn't want one user overriding another's settings and obviously wouldn't change the current default of level 1.

One of the best decisions I have made is make Dozzle as simple as possible. I try to avoid introducing just yet another configuration because it makes my testing harder and it forces me to have different states which isn't ideal. What I would really like to do is just have all these configuration done on the UI or have better defaults. 

My preferred approach would be to choose which settings are type 1 vs other settings of type 2. For example, common searches like #2036 and options like show stopped containers, or automatically redirecting could be relabled as global options. Then display options are just scoped to the browser. 

I personally like this idea more because it's better defaults out of the box and the user doesn't have to have the mental model of what the three scopes actually mean. I think it should be behind the scences, especially for the less technical folks. 

What do you think? And a more important question, when switching computers, which settings would you expect to carry over?

**Comment by EDIflyer** [comment]
Thanks for the reply, Amir!

Totally agree re simplicity - that was actually why I was thinking of it just being an 'all or nothing' setting - either they are applied on a per-browser basis or they are applied on a per-deployment basis (if the appropriate flag is set).  Agree a small JSON may be the easiest way to store things in the container rather than an embedded database.

Personally I wouldn't have any settings that I'd want to be different on different computers/browsers, so if there were global vs local settings I'd still end up having to set some locally.  I also wasn't sure how global options would be for those that have more than one person using it - for example some might want their own searches or showing stopped containers, etc.  I guess the problem is I know only know my use case well and I don't know so much what others do 🙂  You definitely don't want the complexity of handling different users, etc. that's why I was hoping mine was a simpler global solution but very much appreciate it may be different for others! I guess with this only being an option (to make everything global - and perhaps the flag should be `--global_settings` rather than `--server_settings`!) then it at least wouldn't be a regression for those happy with the current setup.

Hope that helps - apologies if it doesn't!!

**Comment by amir20** [comment]
Hmmmmmmmm I think simplest and most intuitive solution is to support multiple users. :) 

Maybe something like `--protected` which would ask to create a new user if one doesn't exist. `adguard/adguardhome` does this. Then the first user can create other users. 

If `--protected` doesn't exist I would just fallback to `localstorage`. 

This could be a lot of work to build. But maybe the most intuitive. The other downside is it would break everybody who already has a username and password. 

But this way would make sense the most because if I know who the user is then I can store data for them. It also makes sense between computers because users expect data to be synced.  

What do you think? It would be a step closer to supporting more use cases too.

**Comment by EDIflyer** [comment]
Oh I mean having settings on a per user basis would certainly solve things from my perspective (I'd just login to each browser with the same credentials), I just thought it might be way way more complex! It would have the advantage of addressing other folks multiple user cases though at the same time.

I've got mine behind Authelia so I wouldn't want to have to login each time, presumably the idea is it could be set to remember who the user is?

**Comment by amir20** [comment]
Ah interesting. That could make it easier for me. 

I could just introduce an `--authenticator`  flag with `simple`, `reverse-proxy` and `none`. When set to reverse proxy I would just rely on something like Authelia to pass me the username and password and not deal with authentication. Then the rest is easy. 

I'll have to research how to get the HTTP header from Authelia.

**Comment by EDIflyer** [comment]
Hopefully this is the sort of thing you're looking for? https://www.authelia.com/integration/trusted-header-sso/introduction/  Obviously just let me know if/when there's anything you want tested!  I've just setup Authelia and Dozzle again on a spare server for playing about with Webtop so have a non-prod server I can easily mess about with.

**Comment by amir20** [comment]
Yes. That's what I found too. But I need a favor. Maybe I am slow or just don't have time with kids. I wasn't able to use the examples to start a simple local instance. I keep getting 404. I spent an hour probably. 

Do you have a compose file with traffek, authorial and example configuration? That would be very helpful if you do. Authelia example folder doesn't seem to work. I am on MacOS.

**Comment by EDIflyer** [comment]
@amir20 ah no probs - I use NPM rather than Traefik but I've got some documentation I've written for the setup if that would be of assistance to you?  Have a look at https://docs.alanjrobertson.co.uk/server-setup/authelia_setup/ and let me know if useful!

**Comment by amir20** [comment]
I'll give this a try. So many config files. I wish it was just one command to bring up a proxy to test. kinda like Dozzle 😬

**Comment by EDIflyer** [comment]
Haha yes, definitely not as simple as Dozzle I'm afraid!! Took me ages to try and get my head round it but in fairness it seems to work very reliably - you can specify bypass, one-factor or two-factor (just remember to restart the container if you change config) and then it checks every request that goes past. I find it a nice-way of putting an authentication barrier in front of multiple services, esp those that don't have their own one built in.

**Comment by amir20** [comment]
Hmm I didn't have luck. Seems like I have to test everything with https and I am trying to test locally without setting up https. 

I am going to try https://github.com/authelia/authelia/tree/master/examples/compose/local to see why it isn't working.

**Comment by amir20** [comment]
Yey finally it worked. Those example are outdated. It took me a while. Now I see

```
Hostname: 06b914b9c021
IP: 127.0.0.1
IP: 172.21.0.3
RemoteAddr: 172.21.0.5:53626
GET / HTTP/1.1
Host: secure.example.com
User-Agent: Mozilla/5.0 (Macintosh; Intel Mac OS X 10_15_7) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/118.0.0.0 Safari/537.36
Accept: text/html,application/xhtml+xml,application/xml;q=0.9,image/avif,image/webp,image/apng,*/*;q=0.8,application/signed-exchange;v=b3;q=0.7
Accept-Encoding: gzip, deflate, br
Accept-Language: en-US,en;q=0.9
Cookie: authelia_session=cu^LJ4Q7lDE62flNdG$S3!4lETNx!U5n
Dnt: 1
Referer: https://authelia.example.com/
Remote-Email: admin@example.com
Remote-Groups: admins,dev
Remote-Name: Admin
Remote-User: admin
Sec-Ch-Ua: "Not=A?Brand";v="99", "Chromium";v="118"
Sec-Ch-Ua-Mobile: ?0
Sec-Ch-Ua-Platform: "macOS"
Sec-Fetch-Dest: document
Sec-Fetch-Mode: navigate
Sec-Fetch-Site: same-site
Sec-Fetch-User: ?1
Upgrade-Insecure-Requests: 1
X-Forwarded-For: 192.168.65.1
X-Forwarded-Host: secure.example.com
X-Forwarded-Port: 443
X-Forwarded-Proto: https
X-Forwarded-Server: a21eaf4c0326
X-Real-Ip: 192.168.65.1
```

Now we are in business. I can assume these headers while testing. I was thinking of using `Remote-Email` because then I can also show user's icon if they have one. 

What do you think?

**Comment by amir20** [comment]
On a different note, I can't believe it took me this long lol.

**Comment by EDIflyer** [comment]
TBH I'm pretty impressed how quickly you sorted it - hope my docs helped to some degree (just let me know of any errors I need to fix BTW!) and definitely quicker than I managed!!

Interestingly `Remote-Name` didn't seem to pass across on mine, but `Remote-Email` and `Remote-User` both fine. I guess it depends what you're planning to use at your end for holding user info - would it just be UN/PWD or would you capture email too?

**Comment by amir20** [comment]
Oh interesting. I didn't notice that. I don't need password any more. I could use email if it exists otherwise fall back to username. Haven't gotten there yet. 

Now comes the hard part of storing all the data without using a database. I'll try to work on it this week and send some incremental updates for you to test.

**Comment by amir20** [comment]
Try testing https://github.com/amir20/dozzle/pull/2442

I just want to make sure the header part works. You should see your profile on the top right like the screenshot.

**Comment by EDIflyer** [comment]
Success!
![image](https://github.com/amir20/dozzle/assets/13610277/06f698fc-2070-48c7-a67e-6f80e84d805b)

**Comment by amir20** [comment]
@EDIflyer while writing the docs, I noticed something. 

I noticed that Authelia uses `Remote-Email`. But Authentik uses `X-authentik-email` and others might use `X-Forwarded-email`. There isn't one accepted way to do this which means each proxy might do different things. 

So I need a way to make this easy in Dozzle. I think I have a few options: 

1. `--authprovider` can be something like`autheliea` or `nginx` which depending on this value will use the right header. I don't like this option because it is very hard to maintain. 
2. Keep as is with value `forward-proxy` and just try all options. Like `Remote-Email` if it exists, then fall back to `X-authentik-email` and finally whatever else I can find. 
3. Last option, make it customiazable with something like `--auth-headers email:X-authentik-email, username:X-authentik-username`. This would add unnecessary configuration which I am not a big fan. 


What do you think? Right now I am leaning on Option 2 because I think it would work for most. But not sure.

**Comment by EDIflyer** [comment]
I agree option 2 sounds sensible given they're clearly quite different options that shouldn't co-exist so iterating through them seems like it should work well. The Authentik ones sounds very specific to them, I wonder if others follow more of the Authelia model?

**Comment by amir20** [comment]
Thanks. I found that Authentik let's you [change](https://goauthentik.io/docs/providers/proxy/custom_headers) the header.  So I am going to go with option 2 and only support for Authelia supports now.

**Comment by amir20** [comment]
Docs published at https://dozzle.dev/guide/authentication

Going to release a version. Please test that to make sure everything is good. Then we can move on to testing #2445 for testing.

**Comment by EDIflyer** [comment]
Interesting - with 5.1.0 not loading (blank page) and is showing an error in the logs:
```
level=info msg="Dozzle version v5.1.0"
level=info msg="Connected to 1 Docker Engine(s)"
level=info msg="Accepting connections on :8080"
level=error msg="Unable to find remote email. Please check your proxy configuration. Expecting header 'Remote-Email'"
level=error msg="Unable to find remote email. Please check your proxy configuration. Expecting header 'Remote-Email'"
```

However I'm passing remote email OK (as you know on the prev test version it looked up Gravatar OK for it)

**Comment by amir20** [comment]
Hmm....I see it too. Let me test a little more.

**Comment by amir20** [comment]
Can you share your compose file? There is a dash now `--auth-provider`. Aside from that I haven't changed anything.

**Comment by EDIflyer** [comment]
```yaml
version: "3"
services:
  dozzle:
    container_name: dozzle
    entrypoint:
      - /dozzle

    environment:
      - PATH=/bin
      - TZ=Europe/London
      - DOZZLE_AUTH_PROVIDER=forward-proxy
    expose:
      - 8080/tcp
    image: amir20/dozzle:latest
    networks:
      - nginx-proxy-manager_default
    volumes:
      - /var/run/docker.sock:/var/run/docker.sock
    working_dir: /
    restart: unless-stopped    
networks:
  nginx-proxy-manager_default:
    external: true
    name: nginx-proxy-manager_default
```
Worked OK with PR2442

**Comment by amir20** [comment]
Hm before I was testing with just Nginx. Now I brought up Authelia and it still works. 

Does #2445 also not work for you? I can add debugging logs there.

**Comment by amir20** [comment]
Added 3425f93b. 

Can you try with `DOZZLE_LEVEL: debug` and running `amir20/dozzle:pr-2445`?

**Comment by amir20** [comment]
@EDIflyer how are we looking? looks like based on your last email things are ok? 

Also, note that #2445 does save to disk. Can you test it?

**Comment by EDIflyer** [comment]
Yes have tried a few times today and seems OK - will check the save to disk in an hour or so, I just need to get home then will map a volume 👍

**Comment by amir20** [comment]
I have released the feature so going to close this now. Next version will have support for `user.yml` and then I will deprecate `--username/password` flags.

**Comment by amir20** [comment]
@EDIflyer can you test `amir20/dozzle:master`? I have merged the other section and want to make sure I didn't break anything. Doing something on my side looks good.

**Comment by EDIflyer** [comment]
@amir20 the only difference I noticed is that `dozzle:master` still shows my Gravatar but doesn't show the email address next to it. dozzle:pr-2445` still shows the email address.

**Comment by amir20** [comment]
I was able to reproduce. Should be fixed now with c74384d0. I guess in some instances it's possible for name to be `''`.
